UI-Service: Proposal for implementation of modals, 3rd iteration

wanze opened the pull request:
Final iteration for the implementation of modals in the UI-Service.
Note: Previous pull-request (https://github.com/ILIAS-eLearning/ILIAS/pull/404) is closed and replace by this one.

wanze committed:
UI-Service: Proposal for implementation of modals

wanze committed:
Save progress on modal implementation (far from complete!)

wanze committed:
Introduction of connector component (which probably shouldn’t be a component since it’s not a “traditional” UI component)

wanze committed:
Refactoring of modals:
Interruptive modals need a title and string message and optionally items that are affected by the destructive operation.
Roundtrip modals accept the content as array
Lightbox modals display their content with lightbox page(s)

wanze committed:
Refactoring of tests

wanze committed:
Remove uncommented code

wanze committed:
Indentation fixes ;-)

Commit:
- Remove connector stuff and action stuff
- Add new interfaces with a first (basic) implementation for the new concept
- Update examples

wanze committed:
Implement a basic lightbox functionality with the help of bootstraps carousel component

Commit:
- Use a javascript module (il.UI.modal) which wraps bootstraps modal functionalities
- Introduce options for signals that are passed to the event handler when the signal is triggered
- Add support to load a modals content via ajax before showing the modal
- Add example how to display a (legacy) form in a modal
- Various updates on examples

Commit:
Fix and update unit tests

Commit:
Improvements on documentation

Commit:
Fix AbstractRendererTest

Amstutz commented on src/UI/Component/Clickable.php:
```diff
@@ -0,0 +1,35 @@
+<?php
+namespace ILIAS\UI\Component;
```
To be discussed: Not sure if I like the path in which those interfaces are stored. I am afraid that we will create a huge mess if we continue to spam the component directory with all sorts of uncategorized interfaces.

wanze replied:
The following interfaces are moved to the implementation folder since they are considered to be "implementation detail" and not relevant for a consumer of the framework:
* TriggeredSignal
* SignalGenerator

wanze replied:
Done: 86acdd74511f4156265146d91ff7751ad0272538

Amstutz commented on src/UI/Component/Clickable.php:
```diff
@@ -0,0 +1,35 @@
+
+	/**
+	 * Get a component like this, triggering a signal of another component on click.
+	 * In contrast to withOnClick, the signal is appended to existing signals for the click event
```
Add . at the end of the sentence.

Amstutz commented on src/UI/Component/Hoverable.php:
```diff
@@ -0,0 +1,35 @@
+
+	/**
+	 * Get a component like this, triggering a signal of another component on hover.
+	 * In contrast to withOnHover, the signal is appended to existing signals for the hover event
```
add .

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 * description:
+	 *   purpose: >
+	 *      An Interruptive Modal disrupts the user in critical situation,
+     *      forcing him or her to focus on the task at hand.
```
intendation

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *        Due to the heavily disruptive nature of this type of Modal it MUST be
+	 *        restricted to critical situations (e.g. loss of data).
+	 *     2: >
+	 *        All delete-actions are critical situations and SHOULD be implemented as an Interruptive Modal.
```
There has been some discussion, whether this rule is to harsh. I propose to add some clarification: "Exceptions are possible if items from lists in forms are to be deleted or if the modal would heavily disrupt the workflow."

klees replied:
I would suggest to add the word "persistent" somewhere. Or "delete from the system" maybe.

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *
+	 * @return InterruptiveItem
+	 */
+	public function interruptiveItem($id, $title, $description = '');
```
The interruptive item should match the items that can be listed in the confirmation screen GUI. This includes an Icon (Type) and a title (see ilConfirmationTableGUI).

wanze replied:
My proposal would be to accept an optional icon in form of an Image object of the UI framework. We could then write an adapter for an ILIAS object to make it easier for the devs (and also offer a factory method), e.g:
```php
$course = new ilObjCourse(123);
$item = $factory->interruptiveILIASItem($course);
```

Amstutz replied:
I would not do this, such strong coupling to ILIAS Objects should IMHO not be part of the UI-Components. All we need ist the possibility to display an icon and title.

wanze replied:
Implemented with 98bb7f5f72d070a574062b37384cf4f66e2438b5
One question we should discuss is how the items are rendered. Currently, they are rendered with an unordered list (already part of the UI framework). Having icons, this is no longer possible. Should we introduce a new list component accepting icons?

wanze replied:
Note: Commit 192144fd1f0e040cadc0149c3dc15fbc4b67858e now renders the items in a simple table which allows to display icon/title/description grouped together. An example was updated to show interruptive items with a course icon and description.

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *   purpose: >
+	 *     Interruptive items are displayed in an interruptive modal and represent the object(s) being affected
+	 *     by the critical action, e.g. deleting.
+	 *   composition:
```
Please describe, what is this control composed of. In this point you may also add some key-aspects to be aware of, when chosing a title etc...

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *     by the critical action, e.g. deleting.
+	 *   composition:
+	 *   effect:
+	 * rules:
```
Do we need a rule for the title of those items? E.g. should they always match the title of the ILIAS-Objects to be deleted? Is an icon (type) mandatory?

wanze replied:
@Amstutz 
I think that an `InterruptiveItem` should be independent from an ILIAS object. As a developer, I would like to use the modal also for other objects where one can't map title/description/icon to an ILIAS object. What do you think about my proposal of introducing an adapter (2 comments) above. We could then write in the doc that if ILIAS objects are used as interruptive items, one should use `$factory->interruptiveILIASItem($myILIASObject);`. Or if not, that the title/desc/icon must match the ones from the ILIAS object.

Amstutz replied:
I agree with the first part of you answer. Please complete the description and rules in clearly stating what MAY, SHOULD and MUST be provided for the Interruptive Item. E.g. do you think and icon is mandatory or not?

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 * description:
+	 *   purpose: >
+	 *     Used to display an image component inside a lightbox modal
+	 *   composition:
```
Please add a description here.

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *   purpose: >
+	 *     Used to display an image component inside a lightbox modal
+	 *   composition:
+	 *   effect:
```
Please describe effects of those pages (e.g. what happens if images are changed).

wanze replied:
@Amstutz I find it really not easy to write the docs, as I'm never sure under what section this belongs and how technical one can be. For example the lightbox modal: I feel like we should write in the docs of the a lightbox modal somewhere that each supported media type (image, video...) is represented by its own page. But this is too technical right? Then again, the effect of what happens when changing the pages does not belong to a page but to the lightbox modal. It is currently written under "composition", not "effect". I'm confused, please somone enlighten me ;)
```
	 *   purpose: >
	 *     The Lightbox Modal displays media data such as images or videos inside a Lightbox.
	 *   composition: >
	 *     If multiple media data are to be displayed in one Lightbox Modal, they can flipped through.
	 *   effect: >
	 *      Lightbox Modals are activated by clicking the full view glyphicon,
	 *      the title of the object or it's thumbnail.
```

Amstutz replied:
Please do not use probably unclear descriptions of other parts as an excuse to not write any description at all. Obviously purpose, composition and effect are not sharp destinction between completely independent dimensions but rather means to structure text to some degree. In the tutorial section in the readme you find short descriptions of what belongs where. If this is unclear, you are welcome to propose a better description. I agree with you, that the flipped through part, rather belongs to effect. You are welcome to propose such a change.

Every component has some composition, so this should never be left blank. Further many have some possible effects (reactions on interactions) that should be described.

Amstutz commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,161 @@
+	 *   composition:
+	 *   effect:
+	 * rules:
+	 * ---
```
are there any restrictions on the images to be displayed? What about title and length of description?

Amstutz commented on src/UI/Component/SignalGenerator.php:
```diff
@@ -0,0 +1,18 @@
+namespace ILIAS\UI\Component;
+
+/**
+ * Interface SignalGenerator
```
At some place we should describe, what a signal is and how it works. Maybe this could be the place.

wanze replied:
Signals are now abstracted with an interface, I added a description there: https://github.com/studer-raimann/ILIAS/blob/86acdd74511f4156265146d91ff7751ad0272538/src/UI/Component/Signal.php

Amstutz commented on src/UI/Component/Triggerable.php:
```diff
@@ -0,0 +1,24 @@
+ * Example: A modal offers signals to show and close the modal.
+ *
+ * A signal is represented by a unique string identifier and may offer some options which can be passed
+ * by a triggerer component when triggering the signal.
```
... or this. Probably forget may comment above about the need of describing the signal, or maybe copy this description.

klees commented on src/UI/Component/Button/Button.php:
```diff
@@ -9,7 +9,7 @@
  * This describes commonalities between standard and primary buttons. 
  */
-interface Button extends \ILIAS\UI\Component\Component, JavaScriptBindable {
+interface Button extends \ILIAS\UI\Component\Component, JavaScriptBindable, \ILIAS\UI\Component\Clickable, \ILIAS\UI\Component\Hoverable {
```
Maybe `use` all this stuff before, will make it more readable imo.

wanze replied:
Jep: f22c6d5046f1fa17387bb81cb087720487995730

klees commented on src/UI/Component/Button/Factory.php:
```diff
@@ -16,7 +16,7 @@
 	 *       is the one that should be used.
 	 *   composition: >
-	 *       The standard button uses the primary bolor as background.
+	 *       The standard button uses the primary color as background.
```
Still nice =)

klees commented on src/UI/Component/Clickable.php:
```diff
@@ -0,0 +1,35 @@
+	 *
+	 * @return $this
+	 */
+	public function appendOnClick($signal, array $options = array());
```
Are we sure, that this is the best name?

wanze replied:
@klees 
We can't use `addOnClick` as this could mean the same as `withOnClick` --> replacing any signal on click. Would you prefer `withAppendOnClick`? My feeling is that this looks strange...

Amstutz commented on src/UI/Component/TriggeredSignal.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Interface TriggeredSignal
+ *
+ * Describes a signal that is triggered by a component on an event
```
Unfortunately, I do not understand this. Why do we need this? Where/when would it be used? Currently, I would need to check in the implementation.

Amstutz commented on src/UI/Component/Triggerer.php:
```diff
@@ -0,0 +1,33 @@
+ *
+ * Describes a component that can trigger signals of other components on given events, such as click or hover.
+ * All supported events are abstracted with interfaces (see Clickable, Hoverable).
+ * Example: A button can trigger the show signal of a modal on click (which will open the modal on button click).
```
Nice, I really like this verbal example in the description. It helps tremendously to quickly grasp the concept.

Amstutz commented on src/UI/Component/Triggerer.php:
```diff
@@ -0,0 +1,33 @@
+
+	const EVENT_CLICK = 'click';
+	const EVENT_HOVER = 'hover';
+	const EVENT_ONLOAD = 'ready';
```
Those three constants seem awkward. Why are they placed here. I think the base interface should be agnostic of it's derivates. For some reason, they are wired up here by constants.

wanze replied:
Constants were removed with 86acdd74511f4156265146d91ff7751ad0272538

Amstutz commented on src/UI/Component/Triggerer.php:
```diff
@@ -0,0 +1,33 @@
+	 *
+	 * @return TriggeredSignal[]
+	 */
+	public function getTriggeredSignals();
```
Now I get the triggered signals...

klees commented on src/UI/Component/Modal/Interruptive.php:
```diff
@@ -0,0 +1,101 @@
+	 *
+	 * @return Interruptive
+	 */
+	public function withTitle($title);
```
Do we have the use case where we change the title of the interruptive modal after creation?

wanze replied:
Not really, you are right. I removed the `with*` methods you mentioned in your comment with the following commit: 095f730e7ff2248abbb66f76b671a207f8af35fd

klees commented on src/UI/Component/Modal/Interruptive.php:
```diff
@@ -0,0 +1,101 @@
+	 */
+	public function withMessage($message);
+
+
```
Do we have the use case where we change the message of the interruptive modal after creation?

klees commented on src/UI/Component/Modal/InterruptiveItem.php:
```diff
@@ -0,0 +1,34 @@
+interface InterruptiveItem {
+
+	/**
+	 * Return a unique ID of the item
```
Cannot guarantee uniqueness. This is just the string that was passed before.

klees commented on src/UI/Component/Modal/Lightbox.php:
```diff
@@ -0,0 +1,31 @@
+	 *
+	 * @return Lightbox
+	 */
+	public function withPages(array $pages);
```
I don't think we need that. Why would I want to change the pages when all a lightbox is is a collection of pages?

klees commented on src/UI/Component/Modal/RoundTrip.php:
```diff
@@ -0,0 +1,70 @@
+	 * @param string $title
+	 * @return RoundTrip
+	 */
+	public function withTitle($title);
```
Do we have a use case for this? I would remove it.

klees commented on src/UI/Component/Modal/RoundTrip.php:
```diff
@@ -0,0 +1,70 @@
+	 * @param \ILIAS\UI\Component\Component|\ILIAS\UI\Component\Component[] $content
+	 * @return RoundTrip
+	 */
+	public function withContent($content);
```
Do we have a use case for this? I would remove it.

klees commented on src/UI/Component/Onloadable.php:
```diff
@@ -0,0 +1,34 @@
+	 *
+	 * @return $this
+	 */
+	public function withOnLoad($signal, array $options = array());
```
Could we have a typehint on some signal-interface or -class here?

klees replied:
This applies to the other On*-interfaces as well.

klees replied:
The `$signal` can't be an arbitrary string, it must be one that was created by some component. So we should hide the underlying type as an implementation detail here.

wanze replied:
Signals are now abstracted via `Signal` interface, see 86acdd74511f4156265146d91ff7751ad0272538

klees commented on src/UI/Component/SignalGenerator.php:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @package ILIAS\UI\Component
+ */
+interface SignalGenerator {
```
I think this is a detail and should go to the implementation folder.

klees commented on src/UI/Component/SignalGenerator.php:
```diff
@@ -0,0 +1,18 @@
+	/**
+	 * Create a unique signal
+	 *
+	 * @return string
```
If we had a signal-interface/-class it could be returned here.

klees commented on src/UI/Component/TriggeredSignal.php:
```diff
@@ -0,0 +1,33 @@
+	 *
+	 * @return array
+	 */
+	public function getSignalOptions();
```
I don't think we need this atm. I'm not sure we need/want this ever.

klees commented on src/UI/Component/Triggerer.php:
```diff
@@ -0,0 +1,33 @@
+
+	const EVENT_CLICK = 'click';
+	const EVENT_HOVER = 'hover';
+	const EVENT_ONLOAD = 'ready';
```
These are implementation details and not part of the public interface imo.

klees commented on src/UI/Implementation/Component/Button/Renderer.php:
```diff
@@ -53,22 +53,29 @@ protected function render_button(Component\Component $component, RendererInterfa
-
-		return $tpl->get();
+		$this->maybeRenderId($component, $tpl);
+        return $tpl->get();
```
Indentation.

klees commented on src/UI/Implementation/Component/Button/Renderer.php:
```diff
@@ -53,22 +53,29 @@ protected function render_button(Component\Component $component, RendererInterfa
 	}
 
-	protected function maybe_render_id($component, $tpl) {
+	protected function maybeRenderId(Component\Component $component, $tpl) {
```
Thx! :+1:

klees commented on src/UI/Implementation/Component/Button/Renderer.php:
```diff
@@ -53,22 +53,29 @@ protected function render_button(Component\Component $component, RendererInterfa
+				$id = $this->createId();
+				$this->triggerRegisteredSignals($component, $id);
+			}
+		}
```
Maybe move this to some place where one could reuse this snippet. Maybe merge this somehow with bindJavascript, e.g. by adding an onload-code?

klees replied:
You could call $this->triggerRegisteredSignals before bindJavascript to do that, I think.

wanze replied:
`triggerRegisteredSignals` is implemented in the `AbstractComponentRenderer` (protected) so it is already available for all Renderers.

klees commented on src/UI/Implementation/Component/Modal/Renderer.php:
```diff
@@ -0,0 +1,182 @@
+	 */
+//	protected function getCancelButton($txt_key) {
+//		return $this->getUIFactory()->button()->standard($this->txt($txt_key), '');
+//	}
```
No uncommented code please.

klees reviewed (commented):
Some nitpicking and questions, as usual. LGTM in general.

klees commented on src/UI/Renderer.php:
```diff
@@ -10,12 +10,14 @@
-	 * @param	$component			Component
-	 * @throws	\LogicException		if renderer is called with a component it can't render
-	 * @return	string
+	 * @return string|array The output of the renderer rendering the component
```
I could live with the additional input type, but the additional output type is imho bad. Could we have another method for that purpose?

wanze replied:
@klees So if multiple components are passed as array, the method returns a concatenated string of each renderer, right? This would also simplify my examples 👍  :)
I guess we do not need another method then?

wanze replied:
Output is now a concatenated string of rendered components, if multiple are passed: f22c6d5046f1fa17387bb81cb087720487995730

klees commented on src/UI/examples/Image/licence_images.txt:
```diff
@@ -0,0 +1,2 @@
+All images taken from https://stocksnap.io
+All photos on StockSnap fall under the Creative Commons CC0 license. That means you can copy, modify, distribute any photo on the site, even for commercial purposes, all without asking permission!
```
I think we need to show this message in the UI of the kitchen sink. Is this implemented?

wanze replied:
Not currently, where should we display this message? As for the modals, I could mention the source of the image in the description of the lightbox page.

klees replied:
Good for me.

Commit:
- Move interfaces “SignalGenerator” and “TriggeredSignal” to implementation folder.
- Use an interface “Signal” to describe the signals instead of a plain string
- Get rid of signal options

Commit:
Modals: Remove unnecessary with* methods

Commit:
Various updates from feedbacks of pull request 421

Commit:
Update examples

Commit:
Update unit tests

Commit:
Add an optional icon to interruptive items

Commit:
Updates on documentation of modal factory, examples and unit tests

Amstutz commented:
Important: This PR will be discussed at the JF on March 13nth. Please leave any feedback until then, so we can have a productive discussion.

chfsx commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,189 @@
+	 *
+	 * @param string $title
+	 * @param string $message
+	 * @param string $form_action
```
form_action = URL?
message = raw string of text or HTML in general?

chfsx commented on src/UI/Component/Modal/Factory.php:
```diff
@@ -0,0 +1,189 @@
+	 *
+	 * @return InterruptiveItem
+	 */
+	public function interruptiveItem($id, $title, Image $icon = null, $description = '');
```
We should introduce a "Icon" soon I think. But for the moment I think Image is ok.

chfsx commented on src/UI/Component/Modal/Lightbox.php:
```diff
@@ -0,0 +1,20 @@
+<?php
+namespace ILIAS\UI\Component\Modal;
+
+use ILIAS\UI\Component\Button;
```
Button is not used here

chfsx commented on src/UI/Implementation/Component/Button/Renderer.php:
```diff
@@ -77,13 +80,13 @@ protected function maybe_render_id($component, $tpl) {
-	 * @inheritdocs
+	 * @inheritdoc
 	 */
 	protected function getComponentInterfaceName() {
```
not part of this but shouldn't this be named getComponentInterfaceName**s** ?

chfsx commented on src/UI/Implementation/Component/Modal/Renderer.php:
```diff
@@ -0,0 +1,187 @@
+		return '';
+	}
+
+	public function registerResources(ResourceRegistry $registry) {
```
missing Doc

chfsx commented on src/UI/Implementation/Component/Modal/Renderer.php:
```diff
@@ -0,0 +1,187 @@
+	/**
+	 * @inheritdoc
+	 */
+	protected function getComponentInterfaceName() {
```
getComponentInterfaceName**s**?

chfsx reviewed (commented):
Great work, especially the whole JavaScript thing!

Commit:
Minor updates after new feedback on pull request

Commit:
Fix JS error when event.preventDefault() is not available
